File: arrays_and_strings/88-merge_sorted_array.cpp
#include <iostream>
#include <vector>

using namespace std;

/*
You're given two integer arrays `nums1` and `nums2`, sorted in non-decreasing
order, and two integers m and n, representing the number of elements
in `nums1` and `nums2` respectively.

Merge `nums1` and `nums2` into a single array sorted in non-decreasing
order.

The final sorted array should not be returned by the function, but instead be
stored inside the array `nums1`. To accomodate this, `nums1` has a length of
`m+n`, where te first `m` elements denote the elements that should be merged,
and the last `n` elements are set to `0` and should be ignored. `nums2` has a
length of `n`.

**Example 1**
Input: nums1 = [1,2,3,0,0,0], m = 3, nums2 = [2,5,6], n =3
Output: [1,2,2,3,5,6]
Explanation: The arrays we are merging are [1,2,3] amd [2,5,6].
The result of the merge is [1,2,2,3,5,6] with the underlined elements coming
from nums1.

 */
class Solution {
public:
  void merge(vector<int> &nums1, int m, vector<int> &nums2, int n) {
    if (m + n < nums1.size())
      return;

    int i = m - 1;
    int j = n - 1;
    int k = m + n - 1;
    while (j >= 0) {
      if (i >= 0 && nums1[i] > nums2[j]) {
        nums1[k] = nums1[i];
        i -= 1;
      } else {
        nums1[k] = nums2[j];
        j -= 1;
      }
      k -= 1;
    }

    for (int num : nums1) {
      cout << num << " ";
    }
  }
};

int main() {
  vector<int> nums1 = {1, 2, 3, 0, 0, 0, 0, 0};
  vector<int> nums2 = {2, 3, 4, 5, 6};
  int m = 3;
  int n = nums2.size();
  Solution().merge(nums1, m, nums2, n);
  return 0;
}

File: arrays_and_strings/88-merge_sorted_array.py
from typing import List


class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        i = m - 1
        j = n - 1
        k = m + n - 1

        while j >= 0:
            if i >= 0 and nums1[i] > nums2[j]:
                nums1[k] = nums1[i]
                i -= 1
            else:
                nums1[k] = nums2[j]
                j -= 1
            k -= 1
        print(nums1)


nums1: List[int] = [0, 1, 2, 3, 4, 0, 0, 0, 0, 0]
nums2: List[int] = [2, 3, 4, 5, 6]
m = 5
n = len(nums2)

Solution().merge(nums1, m, nums2, n)
